### PR TITLE
Full refactor

### PR DESCRIPTION
**Pull request by aslafy-z**
Fully refactored

- Add `file` protocol
- Change syntax to: `git+[repo]//[path]?ref=[ref]` (`:` was allowed in URLs)
- Accepts direct path to Chart (with `Chart.yaml`)
- Fixed bugs
- Still fully POSIX sh compliant
- Add tests
- Setup CI

### CODE REVIEW

**Comment by osterman** [comment]
@aslafy-z what about using the URI format that terraform uses?

E.g. http notation
```
git::https://host.com/org/repo.git?ref=tags/1.9.2
```

E.g. ssh notation
```
git::git@host.com:org/repo.git?ref=branch
```

**Comment by osterman** [comment]
It also supports subpaths with something like...

```
git::git@host.com:org/repo.git//some/path?ref=branch
```

**Comment by osterman** [comment]
Okay, ignore the URL formats above for terraform.

The part I'm not jazzed about is the `@` for the ref since in a URL that typically denotes the basic auth.  Forgive me though, is there prior art for this convention to denote refs with the `@`? If not, perhaps use a query string format since that doesn't limit us to tokens/delimiters. E.g. we can add other parameters down the road.

**Comment by mumoshu** [comment]
But I definitely love conforming to terraform's well defined source URL spec, especially for the path and the branch specification part: `//some/path?ref=branch`

**Comment by mumoshu** [comment]
Oops, we missed each other :)

Yep, it would be even nicer we could find something more typical(in case `@` or `:` aren't typical for denoting the branch and the path, which I believe the case)

**Comment by osterman** [comment]
```
git+protocol://host.com/org/repo//some/path?ref=branch
```

```
git+protocol://host.com/org/repo?ref=branch&path=/some/path
```

Where `protocol` is something like `file` or `ssh` or `https` and `ref` could be `tags/1.2.3` or a branch like `master.

A private repo could be accessed by:
```
git+protocol://user:pass@host.com/org/repo?ref=branch&path=/some/path
```

Or on a non-standard port, by
A private repo could be accessed by:
```
git+https://user:pass@host.com:8123/org/repo?ref=branch&path=/some/path
```

**Comment by aslafy-z** [comment]
Let's use `//path` notation for path, as helm uses an URI parser that skips `#` and `?` parameters. I'll then complete the URI correctly:

```
git+protocol://host.com/org/repo//some/path?ref=branch
git+protocol://host.com/org/repo//some/chart/path/index.yaml?ref=branch
```

File: README.md [hunk 1]
```diff
@@ -55,7 +56,7 @@ You can use it as any other Helm chart repository. Try:
 Fetching also works:
 
     $ helm fetch cert-manager/cert-manager
-    $ helm fetch git+https://github.com/jetstack/cert-manager:release-0.5/contrib/charts/cert-manager-0.5.2.tgz
+    $ helm fetch git+https://github.com/jetstack/cert-manager@release-0.5/contrib/charts/cert-manager-0.5.2.tgz
```
[the remark this answers is not in the document]
Comment (aslafy-z) [reply README.md]: It should already be supported (reference the version within the fetch flag)

**Comment by aslafy-z** [comment]
It should be good by now, anything else you see?

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mumoshu) [reply README.md]: Sorry it should be `0.5.2`(typo!

[the remark this answers is not in the document]
Comment (mumoshu) [reply README.md]: > It should already be supported (reference the version within the fetch flag)

Awesome!!

**Comment by mumoshu** [comment]
Looks good to me!

**Pull request merged**